[test_attention.py] | (T3K) T3000 integration tests | t3000-integration-tests / t3k_motif_tests [wh_llmbox]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] RingJointSDPA TT_FATAL: Non-sliding RingJointSDPA supports batch size 1, got B=2 in t3k_motif_tests
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** (T3K) T3000 integration tests
**Job:** t3000-integration-tests / t3k_motif_tests [wh_llmbox]

### Regression timeline

- **Last passing:** commit `82ef8703abb8` on 2026-08-03T09:00:56Z — https://github.com/tenstorrent/tt-metal/actions/runs/30799629959
- **First failing:** commit `6e06c4479f0a` on 2026-08-03T16:45:41Z — https://github.com/tenstorrent/tt-metal/actions/runs/30833630642/job/91755844223
- **Most recent failing:** commit `0ee137439574` on 2026-08-04T00:21:02Z — https://github.com/tenstorrent/tt-metal/actions/runs/30865291704/job/91858637348

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/30865291704/job/91858637348
- https://github.com/tenstorrent/tt-metal/actions/runs/30833630642/job/91755844223

### Failing test path(s)

- `models/tt_dit/tests/blocks/test_attention.py::test_attention_motif`

All 8 parameterized variants with multi-chip (sequence-parallel) configurations using batch size > 1 fail:
- `test_attention_motif[wormhole_b0-device_params0-motif-2x2sp0tp1]`
- `test_attention_motif[wormhole_b0-device_params0-motif-2x2sp1tp0]`
- `test_attention_motif[wormhole_b0-device_params0-motif-2x4sp0tp1]`
- `test_attention_motif[wormhole_b0-device_params0-motif-2x4sp1tp0]`
- `test_attention_motif[wormhole_b0-device_params0-motif_context_pre_only-2x2sp0tp1]`
- `test_attention_motif[wormhole_b0-device_params0-motif_context_pre_only-2x2sp1tp0]`
- `test_attention_motif[wormhole_b0-device_params0-motif_context_pre_only-2x4sp0tp1]`
- `test_attention_motif[wormhole_b0-device_params0-motif_context_pre_only-2x4sp1tp0]`

### Error excerpt (terminal failure from logs)

```
TT_FATAL: Non-sliding RingJointSDPA supports batch size 1, got B=2  (assert.hpp:104)
##[error]test_attention_motif[wormhole_b0-device_params0-motif-2x2sp0tp1]

RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
info:
Non-sliding RingJointSDPA supports batch size 1, got B=2
```

Full short test summary (same in both runs):
```
FAILED models/tt_dit/tests/blocks/test_attention.py::test_attention_motif[wormhole_b0-device_params0-motif-2x2sp0tp1] - RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
FAILED models/tt_dit/tests/blocks/test_attention.py::test_attention_motif[wormhole_b0-device_params0-motif-2x2sp1tp0] - RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
FAILED models/tt_dit/tests/blocks/test_attention.py::test_attention_motif[wormhole_b0-device_params0-motif-2x4sp0tp1] - RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
FAILED models/tt_dit/tests/blocks/test_attention.py::test_attention_motif[wormhole_b0-device_params0-motif-2x4sp1tp0] - RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
FAILED models/tt_dit/tests/blocks/test_attention.py::test_attention_motif[wormhole_b0-device_params0-motif_context_pre_only-2x2sp0tp1] - RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
FAILED models/tt_dit/tests/blocks/test_attention.py::test_attention_motif[wormhole_b0-device_params0-motif_context_pre_only-2x2sp1tp0] - RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
FAILED models/tt_dit/tests/blocks/test_attention.py::test_attention_motif[wormhole_b0-device_params0-motif_context_pre_only-2x4sp0tp1] - RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
FAILED models/tt_dit/tests/blocks/test_attention.py::test_attention_motif[wormhole_b0-device_params0-motif_context_pre_only-2x4sp1tp0] - RuntimeError: TT_FATAL @ /work/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp:501: B == 1
```

### Reproduction steps

On a T3000 (wh_llmbox) machine with the tt-metal environment set up:

```bash
# Run via the test script
LOG_METAL=run_t3000_motif_tests bash <test_script>

# Or directly via pytest
pytest models/tt_dit/tests/blocks/test_attention.py::test_attention_motif \
  -k 'wormhole_b0 and (2x2sp0tp1 or 2x2sp1tp0 or 2x4sp0tp1 or 2x4sp1tp0)'
```

The failure occurs in all test variants that use sequence-parallel configs with batch size > 1 (B=2). The non-sliding `RingJointSDPA` op asserts `B == 1` at `ring_joint_sdpa_device_operation.cpp:501`.

### Notes

- **Failure class:** `TT_FATAL` assertion — not a timeout or infrastructure flake.
- The failing tests are exclusively those with multi-chip sequence-parallel (SP) configurations (2x2 and 2x4 device meshes). The 1x2 variants (which pass) likely use batch size 1.
- The assertion `B == 1` in `RingJointSDPA::validate_on_program_cache_miss` means the operation does not yet support batch sizes greater than 1 in non-sliding mode. The test fixture is passing B=2.
- The regression was introduced between commits `82ef8703abb8` (passing, 2026-08-03T09:00Z) and `6e06c4479f0a` (failing, 2026-08-03T16:45Z).

**Determinism check:**
- Run 1 (commit `0ee137439574`): `TT_FATAL: Non-sliding RingJointSDPA supports batch size 1, got B=2` at `ring_joint_sdpa_device_operation.cpp:501` (log line 1761)
- Run 2 (commit `6e06c4479f0a`): `TT_FATAL: Non-sliding RingJointSDPA supports batch size 1, got B=2` at `ring_joint_sdpa_device_operation.cpp:501` (log line 1761)
- Both runs: identical 8 FAILED tests, same error class, same source location.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: (T3K) T3000 integration tests`
`Auto-triage-job-name: t3000-integration-tests / t3k_motif_tests [wh_llmbox]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

<!-- PIPELINE-REPORT-START -->
## Pipeline report
Analysis report: https://github.com/tenstorrent/tt-metal/actions/runs/30855158057
Slack notification: https://tenstorrent.slack.com/archives/C0B1F8Z7GQM/p1785793460991409?thread_ts=1785792809.457579&cid=C0B1F8Z7GQM
<!-- PIPELINE-REPORT-END -->

--- a/tests/nightly/t3000/ccl/test_ring_joint_attention.py
+++ b/tests/nightly/t3000/ccl/test_ring_joint_attention.py
@@ -197,6 +197,44 @@ def test_ring_joint_sdpa(
 
 
 @pytest.mark.parametrize(
+    "device_params, all_gather_topology",
+    [
+        (
+            {"worker_l1_size": 1344544, "trace_region_size": 200000, "fabric_config": ttnn.FabricConfig.FABRIC_1D},
+            ttnn.Topology.Linear,
+        ),
+    ],
+    indirect=["device_params"],
+    ids=["line"],
+)
+@pytest.mark.parametrize("mesh_device", [(2, 4)], ids=["2x4"], indirect=True)
+def test_ring_joint_sdpa_multi_batch_wh_t3k(mesh_device, all_gather_topology, reset_seeds):
+    """Non-sliding RingJointSDPA supports B>1 without cross-batch K/V chaining."""
+    submesh = create_ring_joint_sdpa_submesh(mesh_device, rp_axis=0, rp_factor=2, up_axis=1, up_factor=2)
+
+    run_ring_joint_sdpa(
+        submesh,
+        b=2,
+        nh=8,
+        base_seq_len=1024,
+        padded_seq_len=1024,
+        joint_seq_len=0,
+        d=64,
+        q_chunk_size=128,
+        k_chunk_size=128,
+        dtype=ttnn.bfloat16,
+        n_iters=1,
+        trace_enabled=False,
+        num_links=1,
+        rp_axis=0,
+        up_axis=1,
+        all_gather_topology=all_gather_topology,
+        skip_check=False,
+        pcc_threshold=0.994,
+    )
+
+
+@pytest.mark.parametrize(
     "dtype, pcc_threshold",
     [
         (ttnn.bfloat16, 0.994),

--- a/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp
+++ b/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_device_operation.cpp
@@ -497,8 +497,6 @@ void RingJointSDPADeviceOperation::validate_on_program_cache_miss(
             "Chunked GPT-OSS halo {} exceeds the per-device Q slab {}",
             halo_tokens,
             N_local_q);
-    } else {
-        TT_FATAL(B == 1, "Non-sliding RingJointSDPA supports batch size 1, got B={}", B);
     }
 
     TT_FATAL(!(L != 0 && args.is_causal), "Causality is enabled only for ring attention");

--- a/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_program_factory.cpp
+++ b/ttnn/cpp/ttnn/operations/transformer/sdpa/device/ring_joint_sdpa_program_factory.cpp
@@ -893,6 +893,10 @@ tt::tt_metal::ProgramDescriptor build_ring_joint_sdpa_program_descriptor(
     const bool gqa_grouped_kv = ring_joint::is_gqa_grouped_kv_head_mode(v_shares_k_buffer, NH, NHK, NHV);
     const bool k_uses_batch_chain = ring_joint::uses_shared_k_batch_chain(gqa_grouped_kv, NHK);
     const bool use_head_chain = enable_kv_chains && !gqa_grouped_kv;
+    // The store-and-forward chains are scheduled per head, not per (batch, head).
+    // Until their batch-aware scheduling is restored, multi-batch requests read K/V
+    // independently on each core. This preserves the established B>1 functional path.
+    const bool build_kv_chains = enable_kv_chains && B == 1;
 
     const uint32_t q_local_padded_Nt = q_local_padded_N / tt::constants::TILE_HEIGHT;
     const uint32_t kv_local_padded_Nt = kv_local_padded_N / tt::constants::TILE_HEIGHT;
@@ -1968,7 +1972,7 @@ tt::tt_metal::ProgramDescriptor build_ring_joint_sdpa_program_descriptor(
     };
 
     // Build head chains for MHA and separate-V shared-K. GQA uses KV-head-grouped chains instead.
-    if (use_head_chain) {
+    if (use_head_chain && build_kv_chains) {
         for (uint32_t head_id = 0; head_id < static_cast<uint32_t>(head_segments.size()); ++head_id) {
             const auto& segs = head_segments[head_id];
             if (segs.size() < 2) {
@@ -1986,7 +1990,7 @@ tt::tt_metal::ProgramDescriptor build_ring_joint_sdpa_program_descriptor(
 
     // Check query-head chain multicast eligibility and configure mcast for eligible chains.
     uint32_t mcast_chains = 0;
-    if (use_head_chain) {
+    if (use_head_chain && build_kv_chains) {
         struct McastCandidate {
             std::vector<uint32_t> core_indices;
             uint32_t ref_q_chunks;
@@ -2175,7 +2179,7 @@ tt::tt_metal::ProgramDescriptor build_ring_joint_sdpa_program_descriptor(
     bool gqa_mcast_enabled = false;
     std::string gqa_mcast_fallback_reason;
     std::vector<uint32_t> gqa_chain_max_q(gqa_chain_configs.size(), 0);  // per-core loop-padding count
-    if (gqa_grouped_kv && enable_kv_chains) {
+    if (gqa_grouped_kv && build_kv_chains) {
         std::vector<std::vector<ChainSegment>> kv_group_segments(NHK);
 
         for (uint32_t ci = 0; ci < num_cores; ++ci) {
@@ -2261,7 +2265,7 @@ tt::tt_metal::ProgramDescriptor build_ring_joint_sdpa_program_descriptor(
     // Build the shared-K chain for separate-V/latent cases.
     // K is shared across all heads, so all active cores form one chain.
     // Sorted by physical position for a stable unicast ordering (overwritten by mcast pass if eligible).
-    if (k_uses_batch_chain && enable_kv_chains) {
+    if (k_uses_batch_chain && build_kv_chains) {
         std::vector<uint32_t> core_indices;
         for (uint32_t i = 0; i < num_cores; ++i) {
             if (core_work[i].global_q_count == 0) {

